File: code/NewJieliZhiNeng/NewJieliZhiNeng/App设置/关于/DebugView/CustomerCommand/KvoManager.m
//
//  KvoManager.m
//  NewJieliZhiNeng
//
//  Created by EzioChan on 2023/1/6.
//  Copyright © 2023 杰理科技. All rights reserved.
//

#import "KvoManager.h"
#import "JL_RunSDK.h"

@implementation KvoManager

+(instancetype)share{
    static KvoManager *manager;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        manager = [[KvoManager alloc] init];
    });
    return manager;
}



-(void)startListen{
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(notificationAction:) name:nil object:nil];
    //当前连接着的entity
//    JL_EntityM *entity = [[JL_RunSDK sharedMe] mBleEntityM];
    
    //获取真值的方式监听
//    JLModel_Device *dev = [entity.mCmdManager getDeviceModel];
    
//    [dev addObserver:self forKeyPath:@"eqArray" options:NSKeyValueObservingOptionNew context:nil];
//    [dev addObserver:self forKeyPath:@"eqMode" options:NSKeyValueObservingOptionNew context:nil];
//    [dev addObserver:self forKeyPath:@"mAncModeCurrent" options:NSKeyValueObservingOptionNew context:nil];
//
    [JLModel_Device observeModelProperty:@"eqMode" Action:@selector(addEqMode:) Own:self];
    
}

-(void)requestEq{
    //当前连接着的entity
    JL_EntityM *entity = [[JL_RunSDK sharedMe] mBleEntityM];
    [entity.mCmdManager cmdGetSystemInfo:JL_FunctionCodeCOMMON SelectionBit:0x10 Result:^(JL_CMDStatus status, uint8_t sn, NSData * _Nullable data) {
        
    }];
}


-(void)notificationAction:(NSNotification *)note{
    
    if([note.name isEqualToString:@"EQ_MODE_CHANGE"]){
        JL_EntityM *entity = [[JL_RunSDK sharedMe] mBleEntityM];
        JLModel_Device *dev = [entity.mCmdManager outputDeviceModel];
        NSLog(@"dev:%d,dev.eqArray:%@",dev.eqMode,dev.eqArray);
    }

}


-(void)addEqMode:(NSNotification *)note{
    NSLog(@"addNote:%@",note.object);
}




-(void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context{
//    if([keyPath isEqualToString:@"eqArray"]){
//        NSLog(@"%@",change[NSKeyValueChangeNewKey]);
//    }
//    if([keyPath isEqualToString:@"eqMode"]){
//        NSLog(@"%@",change);
//    }
//    if([keyPath isEqualToString:@"mAncModeCurrent"]){
//        NSLog(@"%@",change);
//    }
    JLModel_Device *device = object;
    NSLog(@"device.eqModel:%d",device.eqMode);
    NSLog(@"device.eqArray:%@",device.eqArray);
}



@end

File: code/JieLi_Home_Demo/NewJieliZhiNeng/GlobalViews/AppSettingVC/AboutVC/DebugView/CustomerCommand/KvoManager.swift
import Foundation
import Combine
import JL_BLEKit

class KvoManager: NSObject {
    static let shared = KvoManager()
    private var cancellables = Set<AnyCancellable>()
    
    override private init() {
        super.init()
    }
    
    func startListen() {
        NotificationCenter.default.publisher(for: NSNotification.Name("EQ_MODE_CHANGE"))
            .sink { [weak self] note in
                self?.notificationAction(note)
            }
            .store(in: &cancellables)
        
        JLModel_Device.observeModelProperty("eqMode", action: #selector(addEqMode(_:)), own: self)
    }
    
    func requestEq() {
        guard let entity = JL_RunSDK.sharedMe().mBleEntityM else { return }
        entity.mCmdManager.cmdGetSystemInfo(JL_FunctionCode(rawValue: 0xff)!, selectionBit: 0x10) { status, sn, data in
            
        }
    }
    
    private func notificationAction(_ note: Notification) {
        if note.name.rawValue == "EQ_MODE_CHANGE" {
            guard let entity = JL_RunSDK.sharedMe().mBleEntityM else { return }
            let dev = entity.mCmdManager.outputDeviceModel()
            JLLogManager.logLevel(.DEBUG, content: "dev:\(dev.eqMode),dev.eqArray:\(String(describing: dev.eqArray))")
        }
    }
    
    @objc private func addEqMode(_ note: Notification) {
        if let obj = note.object {
            JLLogManager.logLevel(.DEBUG, content: "addNote:\(obj)")
        }
    }
    
    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        if let device = object as? JLModel_Device {
            JLLogManager.logLevel(.DEBUG, content: "device.eqModel:\(device.eqMode)")
            JLLogManager.logLevel(.DEBUG, content: "device.eqArray:\(String(describing: device.eqArray))")
        }
    }
}
